src/libraries/System.Private.CoreLib/src/System/Text/Unicode/Utf16Utility.Validation.cs

```diff
@@ -87,17 +88,32 @@ static Utf16Utility()
                     Vector128<short> vector8800 = Vector128.Create(unchecked((short)0x8800));
                     Vector128<ushort> vectorZero = Vector128<ushort>.Zero;
 
+
+                    Vector128<byte> initialMask = Vector128.Create((ushort)0x1001).AsByte();
+                    Vector128<byte> mostSignficantBitMask = Vector128.Create((byte)0x80);
                     do
                     {
-                        Vector128<ushort> utf16Data = Sse2.LoadVector128((ushort*)pInputBuffer); // unaligned
+                        Vector128<ushort> utf16Data;
+                        if (AdvSimd.Arm64.IsSupported)
+                        {
+                            utf16Data = AdvSimd.LoadVector128((ushort*)pInputBuffer); // unaligned
+                        }
+                        else
+                        {
+                            utf16Data = Sse2.LoadVector128((ushort*)pInputBuffer); // unaligned
+                        }
                         uint mask;
 
                         Vector128<ushort> charIsNonAscii;
-                        if (Sse41.IsSupported)
-                        {
-                            // Sets the 0x0080 bit of each element in 'charIsNonAscii' if the corresponding
-                            // input was 0x0080 <= [value]. (i.e., [value] is non-ASCII.)
 
+                        // Sets the 0x0080 bit of each element in 'charIsNonAscii' if the corresponding
+                        // input was 0x0080 <= [value]. (i.e., [value] is non-ASCII.)
+                        if (AdvSimd.Arm64.IsSupported)
```

Comment:
I feel that the approach of branching on each individual instruction might limit the flexibility of optimizations that can be applied. I would probably just inline all architecture-specific optimizations in a single block, extracting any reusable code into inlined helper methods.

Reply:
There isn't actually a branch here, it is optimized away entirely. The JIT should generally do a good job of this, but there may be cases where it is less than ideal.

It's probably worth looking and checking. Having larger contiguous blocks would help with readability, however.